train.py

```diff
@@ -288,8 +288,14 @@ def loss_fn(pred, labels):
             else:
                 # Non-PP forward / backward
                 with train_context():
-                    pred = model(input_ids)
-                    loss = loss_fn(pred, labels)
+                    with contextlib.nullcontext() if parallel_dims.dp_shard_enabled else torch.autocast(
+                        "cuda",
+                        dtype=TORCH_DTYPE_MAP[
+                            job_config.training.mixed_precision_param
+                        ],
+                    ):
```

Comment:
IMO there are two ways:
1. To simplify it here, we can have a separate `get_foward_only_context()` which encapsulates the logic.
2. Modify `get_train_context` to receive a boolean variable `is_backward` so that we can obtain two different contexts, one for forward, the other for backward. Pro: only one method to obtain context. Con: need to enter and exit twice for context managers shared by forward and backward.
cc: @awgu wdyt?

Reply:
2 sounds good to me

Reply:
will `maybe_enable_compiled_autograd` work correctly when it exits and re-enters in-between a fwd-bck?

Reply:
@fegin

Reply:
ye, that should work. It's a context manager to enable the flag. So it should be okay to separately wrap fwd and bwd with `maybe_enable_compiled_autograd`.